### DISCUSSION TITLE
Core Data support

### DISCUSSION THREAD

**Discussion by chbeer** [skiptools/skip#124] [Ideas]
Using something like this on Kotlin side:
- Room
- [NexusData](https://github.com/dkharrat/NexusData)
- [ORM Lite](http://ormlite.com/sqlite_java_android_orm.shtml)
- [Green DAO](http://greendao-orm.com/)

**Comment by marcprux** [comment]
We do currently have [SkipSQL](https://skip.tools/docs/modules/skip-sql/) for low-level access to SQLite databases, along with some sample apps like [DataBake](https://skip.tools/docs/samples/skipapp-databake/) that demonstrates how the library can be used in an app.

We don't currently have any higher-level ORM solution. We are considering the best approach. Porting all of Core Data is unlikely for various reasons (such as assumption of an Objective-C environment), but implementing [SwiftData](https://developer.apple.com/xcode/swiftdata/) has been considered, or else a third party solution like [GRDB.swift](https://github.com/groue/GRDB.swift). Those Android libraries you mention are also good suggestions as well – Room is clearly popular, and has many nice features.

We would be interested in hearing the feedback and insights of the community into what features and levels of abstraction they would like to see. Every ORM is a series of trade-offs, as the Skip development teams is acutely aware: we created the original [OpenJPA](https://openjpa.apache.org) project ages ago. So please do use this thread to let us know what you would like, and what you would **not** like, to see in any dual-platform ORM solution we might eventually bring to Skip.

**Reply by aehlke** [reply]
Now that Realm is dead, it would be great to have SwiftData support

**Reply by aehlke** [reply]
But if that's too much work because there's no 1:1 analog (although I bet there are if you constrain models with some data type limitations etc) on Android/elsewhere, it would also be great to have GRDB rather than have to adopt SkipSQL which is a new lib that few have experience with yet. But SwiftData would really be ideal as it's clearly the future (though it's going to take years to mature)

**Reply by marcprux** [reply]
Both [GRDB.swift](https://github.com/groue/GRDB.swift) and [SQLite.swift](https://github.com/stephencelis/SQLite.swift) are excellent projects, and they could probably be adapted to work on top of [SkipSQL](https://github.com/skiptools/skip-sql) with a bit of effort. Interested has been expressed, but we haven't yet seen any community initiative towards this.

**Reply by aehlke** [reply]
I see you've integrated SQlite.swift - nice to see progress there. Unfortunately it's lacking things that GRDB has such as iCloud sync via CKSyncEngine (via Harmony), and change notifications. Its original author recommends GRDB these days, and performance-wise it's the furthest behind GRDB and Lighter-swift

**Reply by marcprux** [reply]
Agreed that SQLite.swift and GRDB are different levels. The SQLite.swift module exists as part of swift-sqlite primarily as a conduit to the native SQLite C build.

We don't have concrete plans yet to work on GRDB, but the native Swift toolchain side might ease the process of getting it working.

**Reply by aehlke** [reply]
Thank you for the update I realized that's why after reading the code. Telegraphing what more productionized DB you might pick first helps with starting projects early before it lands, understanding plans and roadmaps change continuously with the work... Understood there's nothing to update on that but very interested in following updates on whatever tool you choose first (more so than any particular one)

**Reply by aehlke** [reply]
@marcprux I saw your GRDB work is merged/complete. Is there an update on being able to use SQLiteData in Skip? Thank you

**Reply by marcprux** [reply]
Nothing yet. My PR https://github.com/groue/GRDB.swift/pull/1708 was just closed in favor of a strategy to facilitate a fork-friendly setup (in https://github.com/groue/GRDB.swift/pull/1845 and further discussed at https://github.com/groue/GRDB.swift/pull/1827#issuecomment-3761902843). Once that is merged, we would then need to engage with [sqlite-data](https://github.com/pointfreeco/sqlite-data) to encourage them to add support for this strategy, potentially through traits.

So we have a little ways to go, but the groundwork is being laid and I am hopeful we'll have a working solution in the not-too-distant future.

**Comment by vdkdamian** [comment]
I'd really like to see SwiftData support coming.
